hack/codegen.go

```diff
@@ -0,0 +1,734 @@
+					gv, _ := gofunc.Args.Get(goKeys[index])
+
+					if !reflect.DeepEqual(fv, gv) {
+						signatureChanges = append(signatureChanges, SignatureChange{
+							Function: gofunc.Name,
+							Have:     argsFromMap(gofunc.Args, true),
+							Want:     argsFromMap(feature.Args, true),
+						})
+
+						break FeaturesLoop
+					}
+				}
+
+			}
+		}
+	}
+
+	// 10. check signatures are ok
+	if len(signatureChanges) != 0 {
+		var argBuf bytes.Buffer
+		for _, sc := range signatureChanges {
+			argBuf.WriteString(fmt.Sprintf(`
+function %v
+	have %v
+	want %v
+`, sc.Function, sc.Have, sc.Want))
+		}
+
+		return fmt.Errorf("source file %v has a different signature/s:\n %v", mapping.GoFile, argBuf.String())
+	}
+
+	// 11. if in sync, nothing to do
+	if isInSync {
+		return nil
+	}
+
+	// 12. New go feature file
+	if !isGoFileOk {
+		if !update {
+			return fmt.Errorf("generated code is out of date (from %v feature, new go file %v)",
+				mapping.FeatureFile, mapping.GoFile)
+		}
+
+		log.Printf("Generating new go file %v...", mapping.GoFile)
+		buf := bytes.NewBuffer(make([]byte, 0))
+
+		err := template.Execute(buf, mapping)
+		if err != nil {
+			return err
+		}
+
+		// 10. if update is set
+		if update {
+			isDirOk := files.IsDir(mapping.GoFile)
+			if !isDirOk {
+				err := os.MkdirAll(filepath.Dir(mapping.GoFile), 0755)
+				if err != nil {
+					return err
+				}
+			}
+
+			err := ioutil.WriteFile(mapping.GoFile, buf.Bytes(), 0644)
+			if err != nil {
+				return err
+			}
+
+			featFile := filepath.Base(path)
+			log.Printf(`Please add '"features/%v": %v.FeatureContext' to features map defined in conformance_test.go (order matters)`,
+				featFile, mapping.Package)
+		}
+
+		return nil
+	}
+
+	if len(mapping.NewFunctions) == 0 {
+		return nil
+	}
+
+	// 13. if update is set
+	if update {
+		log.Printf("Updating go file %v...", mapping.GoFile)
+		err := updateGoTestFile(mapping.GoFile, mapping.NewFunctions)
+		if err != nil {
+			return err
+		}
+	}
+
+	if len(mapping.NewFunctions) != 0 {
+		return fmt.Errorf("generated code is out of date")
+	}
+
+	return nil
+}
+
+// Function holds the definition of a function in a go file or godog step
+type Function struct {
+	// Name
+	Name string
+	// Expr Regexp to use in godog Step definition
+	Expr string
+	// Args function arguments
+	// k = name of the argument
+	// v = type of the argument
+	Args *orderedmap.OrderedMap
+}
+
+type Mapping struct {
+	Package string
+
+	FeatureFile string
+	Features    []Function
+
+	GoFile        string
+	GoDefinitions []Function
+
+	NewFunctions []Function
+}
+
+// SignatureChange holds information about the definition of a go function
+type SignatureChange struct {
+	Function string
+	Have     string
+	Want     string
+}
+
+var templateFuncs = template.FuncMap{
+	"backticked": func(s string) string {
+		return "`" + s + "`"
+	},
+	"unescape": func(s string) template.HTML {
+		return template.HTML(s)
+	},
+	"argsFromMap": argsFromMap,
+}
+
+const goTemplate = `
```

Comment:
Would it be possible to extract this goTemplate in its own file? I feel like IDE support would be better if it could parse the template rather than seeing it as a string.

Reply:
Fixed. Extracted to `hack/codegen.tmpl`